Display commit author in preview

eseliger opened the pull request:
Will rework the design, once we have the final version. This, in the meantime, should work, so I think it's fine to review.

Works on https://github.com/sourcegraph/sourcegraph/issues/13645

eseliger commented on cmd/frontend/graphqlbackend/person.go:
```diff
@@ -37,52 +37,53 @@ func (r *personResolver) resolveUser(ctx context.Context) (*types.User, error) {
 	}
 	if user != nil && user.Username != "" {
-		return user.DisplayName, nil
+		return user.Username, nil
```
I _think_ this was a bug

mrnugget replied:
Yeah, looks like it.

eseliger commented on cmd/frontend/graphqlbackend/person.go:
```diff
@@ -37,52 +37,53 @@ func (r *personResolver) resolveUser(ctx context.Context) (*types.User, error) {
-func (r *personResolver) DisplayName(ctx context.Context) (string, error) {
+func (r *PersonResolver) DisplayName(ctx context.Context) (string, error) {
 	user, err := r.resolveUser(ctx)
-	if err != nil && !errcode.IsNotFound(err) {
```
these error code checks were unnecessary, resolveUser already does that and resets err to nil

eseliger commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -4555,9 +4570,9 @@ type Person {
     """
     displayName: String!
     """
-    The avatar URL.
+    The avatar URL, if known.
     """
-    avatarURL: String!
+    avatarURL: String
```
Same behavior as before, but now properly documented (only null over ""), but it's not good to put invalid URLs in image sources in the frontend, so made this nullable properly

mrnugget commented on web/src/enterprise/campaigns/apply/VisibleChangesetSpecNode.story.tsx:
```diff
@@ -33,7 +33,19 @@ export const visibleChangesetSpecStories: Record<string, VisibleChangesetSpecFie
-            commits: [{ message: 'Commit message' }],
+            commits: [
+                {
+                    subject: 'Commit message',
```
It's really easy to confuse this with a header when looking at the story in the Storybook.


```suggestion
                    subject: 'This is the first line of the commit message',
```

I always try to use real looking data in my examples vs. data that describes what the data is.

Good:

```go
func add(a, b int) (sum int) { return a + b }
```

Bad:

```go
func function(param1, param2 type) (returnValue type) { return param1 + param2 }
```

You get the idea :)

mrnugget commented on web/src/enterprise/campaigns/apply/VisibleChangesetSpecNode.story.tsx:
```diff
@@ -33,7 +33,19 @@ export const visibleChangesetSpecStories: Record<string, VisibleChangesetSpecFie
+                {
+                    subject: 'Commit message',
+                    body:
+                        'And the more explanatory body. And the more explanatory body. And the more explanatory body. And the more explanatory body. And the more explanatory body. And the more explanatory body. And the more explanatory body. And the more explanatory body. And the more explanatory body. And the more explanatory body. And the more explanatory body. And the more explanatory body.',
```
I suggest using a message here that's wrapped at 80 chars. Or both in the same message. But this is not representative of what most commit messages look like, IMHO.

mrnugget commented on web/src/enterprise/campaigns/apply/VisibleChangesetSpecNode.story.tsx:
```diff
@@ -54,7 +66,19 @@ export const visibleChangesetSpecStories: Record<string, VisibleChangesetSpecFie
-            commits: [{ message: 'Commit message' }],
+            commits: [
+                {
+                    subject: 'Commit message',
```
Same as above.

mrnugget commented on web/src/integration/campaigns.test.ts:
```diff
@@ -546,7 +546,18 @@ describe('Campaigns', () => {
                                             commits: [
                                                 {
-                                                    message: 'Commit message',
+                                                    subject: 'Commit message',
```
Same as above.

mrnugget commented on web/src/enterprise/campaigns/apply/GitBranchChangesetDescriptionInfo.tsx:
```diff
@@ -0,0 +1,68 @@
+                                        className="btn btn-secondary btn-sm px-1 py-0 font-weight-bold align-item-center mr-2"
+                                        onClick={toggleShowCommitMessageBody}
+                                    >
+                                        &#8943;
```
I assume these are the "..."? I'm not sure what the practice in the `./web` is, but I think a _tiny_ comment here is helpful :) Or even `const dotDotDot = "&#8943;"`.

mrnugget commented on web/src/enterprise/campaigns/apply/GitBranchChangesetDescriptionInfo.tsx:
```diff
@@ -0,0 +1,68 @@
+
+    return (
+        <>
+            <h4>Commits</h4>
```
Since we only have single commits (for now and the foreseeable future) per changeset — does it make sense to mention the plural here?

That being said, I think the header itself might be unnecessary.

LawnGnome replied:
Looking at the storybook, I think it's necessary now, although I wonder if we'll end up changing this into a set of tabs rather than displaying the commit and diff in full every time (which would then subsume the headings into the tab bar).

mrnugget commented on cmd/frontend/graphqlbackend/person.go:
```diff
@@ -10,12 +10,12 @@ import (
-	name  string
-	email string
+type PersonResolver struct {
+	UserName  string
```
Regarding snake case: It's a bit weird that it's `UserName` here but `user.Username` further down.

Also, why add the `User`? I think `name` to be returned as default by `Name()` makes sense (same for `email`). Maybe `defaultName` is the better name to make distinguishing between all the names here easier?

eseliger replied:
If I change it to `name`, I need a factory function (because we use the resolver outside of the `graphqlbackend` package. I can certainly do that, though.

LawnGnome replied:
I think it's worth it: `Username` feels confusing to me as well, since it's not _really_ a Sourcegraph username (right?).

mrnugget replied:
`DefaultName` or `FallbackName` would also work, IMHO, and not require the factory function.

mrnugget reviewed (approved):
I think the idea of surfacing this data is really cool, because it makes it easier to understand for users what the `changesetTemplate` does 👍  

That being said: I'm not sure whether it isn't too repetitive to show the same commit message everywhere. We already have that for the pull request titles and that I don't like 100% either 😁 

What are your thoughts here?

And, that being said part 2: with the commit message it might be possible to simplify the design of the preview a bit and get rid of the headers. If we show the commit message we probably don't need to show "Commits" and "Diff" as headers, since the context makes it pretty clear what is what. Maybe we can take some inspiration from GitHub:

![image](https://user-images.githubusercontent.com/1185253/94410746-023a2380-0178-11eb-8d68-1df9bcecc7cc.png)

eseliger commented:
> That being said: I'm not sure whether it isn't too repetitive to show the same commit message everywhere. We already have that for the pull request titles and that I don't like 100% either 😁

I think that is actually fine. It is in the collapsible menu fo a changeset, so the user explicitly asks for further information. For each changeset opened, I would probably like to be able to confirm the impact it has, so showing the (currently just) repeated data feels not too weird to me. Happy to take suggestions though. Also, please note this design is not final :) I just wanted to close the main work here, so I can focus on other things first and do the quick design overhaul once we have the designs.

> And, that being said part 2: with the commit message it might be possible to simplify the design of the preview a bit and get rid of the headers. If we show the commit message we probably don't need to show "Commits" and "Diff" as headers, since the context makes it pretty clear what is what. Maybe we can take some inspiration from GitHub:

< image >

Hmm, I think why GitHub doesn't need it there is because you purposefully went on a page to view a commit there, hence the entity expected is the commit. With out list of changesets, though, I think especially for first-time-users it's not immediately clear what additional information you can get on a changeset, and having those two short guidelines helps grasping that you can get the commits created and the overall diff here in this view. Not sure that makes sense, but that was how I felt about it. Additionally, I'll also talk about this with design and it was just my personal note added here, until design's ready.

LawnGnome commented on web/src/enterprise/campaigns/apply/VisibleChangesetSpecNode.tsx:
```diff
@@ -98,27 +102,34 @@ export const VisibleChangesetSpecNode: React.FunctionComponent<VisibleChangesetS
+                                    description={node.description}
+                                    isExpandedInitially={expandChangesetDescriptions}
+                                />
+                                <h4>Diff</h4>
```
This feels inconsistent, since `GitBranchChangesetDescriptionInfo` has its `h4` built into it: maybe that header should move out here as well?

LawnGnome reviewed (approved):
LGTM. I think the design does need some more work — it feels like it uses more vertical space than is necessary — but let's get the  foundation built before we worry too much about the paint colour. :smiley:

eseliger commented:
Addressed feedback, design is simplified as per a quick design session yesterday, so merging for now so we have the necessary foundation of this feature merged.
